README.md
django-bingo
------------

django-bingo is a bingo app to play bingo games like "bullshit bingo" while listening together to some radio show and marking the words, which were mentioned.

Dependencies
------------

Python packages:

* django 1.8
* django-jquery 1.9.1
* django-colorful 0.1.3
* pytz 2013b
* pillow 2.1.0

For SSE:

* sse 1.2
* redis 2.10.3
* flask 0.10.1
* gevent 1.0.2

Other: a ttf-file as font for the image-export.

Installing
----------

Create a django project. Then add the following options to your settings.py

* append to ```INSTALLED_APPS``` (make sure ``bingo`` is included first)
    * 'bingo'
    * 'jquery'
    * 'colorful'
    * 'registration'
* ```FONT_PATH``` = '/path/to/font.ttf'

Include the bingo-app in your project urls.py, like this:
```url(r'^', include('bingo.urls'))```
Include the ```registration``` app like this:
```url(r'^accounts/', include('registration.backends.simple.urls'))```

Registration is only tested with the ```simple``` backend.
You can use other ```django-registration``` backends, but you will need to add additional templates and views.

You can use the Sites framework to add several different Bingo sites.
Different Sites have different sets of Words, and they will generate a different set of Games, so the Bingos are independend from each other.

Upgrading
---------

For upgrading from an older version you may need to use ```manage.py migrate```, if the database scheme was changed.

When upgrading from versions prior to 1.4, upgrade first using ``south`` and ```django 1.6```. Then upgrade to the current version using the django migrations.

Settings
--------

#### required settings

* Add ``bingo.context_processors.bingo`` to the ``context_processors`` in the [OPTIONS](https://docs.djangoproject.com/en/1.10/ref/settings/#std:setting-TEMPLATES-OPTIONS) section of the ```TEMPLATE``` settings.
* Add at least one of
    * ```GAME_HARD_TIMEOUT``` time after which a game is stopped
    * ```GAME_SOFT_TIMEOUT``` time after which a game without any activity is stopped

#### optional settings

see CONFIGURATION.md

Customizing
-----------

Many minor changes can be done with [django-apptemplates](https://pypi.python.org/pypi/django-apptemplates/). See their documentation for how to add the module to the project.

After installation you can use add additional template folders in ``DIRS`` list of the ``TEMPLATES`` option in ``settings.py`` for customizing your installation. this allows for example a list like ``["thisbingosite_templates", "bingosites_templates", "default_templates"]`` for using template folders shared between different sites.

Example for ``thisbingosite_templates/bingo/main.html``:

    {% extends "bingo:bingo/base.html %}
    {% block extra_content_top %}
	    Welcome to my django-bingo game!
	    {{ block.super }}
	{% endblock %}


ChangeLog
django-bingo 1.7.0 ():
	* new: Tweet Link
	* new: Meta tags for a "Twitter Card", which shows a board thumbnail and
	       game metadata inside tweets with a board link.
	* new: Tag Cloud on the main page and the game page.
	* new: Twitter Card for the Game page (page with all boards of a game).
	* new: Show "create" date of the boards on the Profile page.
	* changed: Words can now be marked as "topic", "meta" and "middle".
	           .is_middle is now replaced by .type == TYPE_MIDDLE.
	* new: The board url is added as a watermark in the lower right of the
	       board images.
	* fixed: The server-sent-events script timeout for inactive connections works
	         now again and disconnects connections to the redis backend correctly.
	* new: GAME_WEEK_DAYS option to limit game start to certain days of week.
    * new: Twitter card for the profile page.
	* fixed: catch Redis connection errors.
	* improved: Adapted to Django 1.10

django-bingo 1.6.0 (2015-09-05):
	* new: Show game description above the bingo board.
	* fix: Disable USE_SSE, if no redis module is found, but USE_SSE is set.
	* improved: Adapted to Django 1.8
	* improved: show 12 boards per game on the front page, so they are aligned
	  for different window sizes with 2, 3, 4, 6 or 12 boards per row.

django-bingo 1.5.1 (2015-06-25):
	* improved: change veto link from [-] to [Veto] on mouseover.
	* improved: use fieldset on forms.
	* fixed: image cache works correct with different sites now.
	* improved: added missing translations

django-bingo 1.5.0 (2014-12-18):
	* improved: header row and number of players column on game list page.
	* new: support for User profiles, with the User's BingoBoards
	  associated to the account. Old BingoBoards can be claimed, if they have
	  a password, which is unique for the old Game.
	* fixed: corrected url regex for numbers, not to match ""
	* fixed: off by one error, causing the bottom/right lines to be 1px too
	  thin.
	* new: optionally use server-sent events for less polling (needs a redis
	  database)
	* improved: added POLLING_INTERVAL / POLLING_INTERVAL_SSE settings to
	  configure how often updated votes / user numbers should be fetched
	* improved: show title text below the board, if javascript is enabled
	* new: the bingo page has now a variable middleword set.
	* improved: Game and BingoBoard now have ``get_absolute_url``, which
	  provides i.e links to the game/bingo page in the admin interface.
	* improved: style fixes ("veto" title tag)
	* improved: added game descriptions to the game list page.

django-bingo 1.4.0 (2014-02-27):
	Attention: when updating from an older version, you will need south
	to migrate the database.
	* new: added a "description" field to Word model, which is shown as
	  title="..." attribute on bingo fields.
	* new: added a "description" field to Game model, which can be set when
	  creating the first board of a game.
	* performance: reduced querys to get the number of votes for a Word in a Game
	* improved: the bingo page is reloaded after the game is expired, so it
	  stops polling for new votes.
	* performance: update the last_used timestamp with .update() instead of
	  .save(), which avoids saving the whole object, when only the timestamp
	  has changed.
	* performance: reduced number of queries when creating board lists.
	* style: some style fixes.
	* new: option to rate the show (not the board!)
	* improved: [-] veto buttons are now toggle buttons (veto/neutral)

django-bingo 1.3.2 (2014-02-20):
	* added: MANIFEST.in to include package_data in source distributions

django-bingo 1.3.1 (2014-01-21):
	* fix: avoid creating duplicate BingoBoards / Games

django-bingo 1.3.0 (2013-11-26):
	* fix: mark "my board" on the game page, too.
	* new: added a light theme, which overrides the dark theme, which can be
	  added to the theme switcher, when the dark theme is the default THEME.
	* improved: title-text for board thumbnails with creation time. alt-text
	  with board_id.
	* improved: board thumbnails switch between voted / marked on mouseover.
	* new: VOTE_START_TIME allows to set a time between GAME_START_TIMES[0]
	  and GAME_END_TIME, after which voting will be activated. This can be
	  used to allow board creation before the game starts, but disable early
	  votes.
	* new: added a footer.

django-bingo 1.2.1 (2013-10-03):
	* fix: each BingoBoard was missing one of the words, which are not on the board.

django-bingo 1.2.0 (2013-09-27):
	* new feature: show boards as thumbnails in the board list.
	* changed: removed is_active. to set a Word to inactive, just remove it from all sites.
	* changed: added "unique" constraint to field Word.word.
	* improved: added some more list-columns and list-filters in the admin interface.
	* new feature: implemented theme support
	  * THEME = "somewhere/myfile.css" is the path for a site-theme
	  * THEMES = (('name', 'path/to/file.css'), ...) are themes shown in a theme chooser
	  * new dark theme included. use "bingo/themes/dark.css" as path.
	* Note: you now need to add TEMPLATE_CONTEXT_PROCESSORS to your config, see the README.

django-bingo 1.1.1 (2013-09-21):
	* fixed: GAME_END_TIME did not stop currently running games

django-bingo 1.1.0 (2013-09-14):
	* performance: optimized queries by using select_related().
	* performance: implemented a vote cache, which avoids a lot of queries.
	* style: use "background-attachment: fixed" for the background image.
	* pep8: urls.py is now pep8 compliant.


bingo/models.py
from django.contrib.auth.models import User
from django.utils.translation import ugettext as _
from django.core.exceptions import ValidationError
from django.conf import settings
from django.utils import timezone
from django.db import models, transaction
from django.db.models import Count, Sum
from django.contrib.sites.models import Site
from django.core.cache import cache
from django.core.urlresolvers import reverse

from colorful.fields import RGBColorField

from random import randint

from times import is_starttime, is_after_endtime, get_endtime


# Color ranges
COLOR_FROM = getattr(settings, "COLOR_FROM", 80)
COLOR_TO = getattr(settings, "COLOR_TO", 160)

# Expire game after hard-timeout seconds, or
# soft-timeout seconds inactivity, whatever comes first.
GAME_SOFT_TIMEOUT = getattr(settings, "GAME_SOFT_TIMEOUT", 60)
GAME_HARD_TIMEOUT = getattr(settings, "GAME_HARD_TIMEOUT", 120)

BINGO_IMAGE_DATETIME_FORMAT = getattr(
    settings, "BINGO_IMAGE_DATETIME_FORMAT", "%Y-%m-%d %H:%M")

# if a user did not vote/refresh his board for this time,
# he is counted as inactive
USER_ACTIVE_TIMEOUT = getattr(settings, "USER_ACTIVE_TIMEOUT", 5)

THUMBNAILS_ENABLED = getattr(settings, "THUMBNAILS_ENABLED", True)

WORD_TYPE_TOPIC = 1
WORD_TYPE_MIDDLE = 2
WORD_TYPE_META = 3
WORD_TYPES = (
    (WORD_TYPE_TOPIC, 'Topic'),
    (WORD_TYPE_MIDDLE, 'Middle'),
    (WORD_TYPE_META, 'Meta'),
)


class Word(models.Model):
    """
        a word entry. Should not be deleted, but only disabled
        to preserve old BingoFields
    """
    word = models.CharField(max_length=255, unique=True)
    description = models.CharField(max_length=255, blank=True)
    type = models.PositiveSmallIntegerField(choices=WORD_TYPES)
    site = models.ManyToManyField(Site, blank=True)

    class Meta:
        ordering = ("word",)

    def __unicode__(self):
        return u"Word: " + self.word


class TimeRangeError(Exception):
    pass


def get_game(site, description="", create=False):
    """
        get the current game, if its still active, else
        creates a new game, if the current time is inside the
        GAME_START_TIMES interval and create=True
        @param create: create a game, if there is no active game
        @returns: None if there is no active Game, and none shoul be
        created or the (new) active Game.
    """

    game = None
    games = Game.objects.filter(site=site).order_by("-created")
    try:
        game = games[0]
    except IndexError:
        game = None

    # no game, yet, or game expired
    if game is None or game.is_expired() or is_after_endtime():
        if create:
            if is_starttime():
                game = Game(site=site, description=description)
                game.save()
            else:
                raise TimeRangeError(
                    _(u"game start outside of the valid timerange"))
        else:
            game = None

    # game exists and its not after the GAME_END_TIME
    elif not is_after_endtime():
        game = games[0]

    return game


class Game(models.Model):
    game_id = models.IntegerField(blank=True, null=True)
    description = models.CharField(max_length=255, blank=True)
    site = models.ForeignKey(Site)
    created = models.DateTimeField(auto_now_add=True)
    last_used = models.DateTimeField(auto_now_add=True)

    class Meta:
        unique_together = ("game_id", "site")

    def get_absolute_url(self):
        return reverse('bingo.views.game', kwargs={"game_id": self.game_id})

    def __unicode__(self):
        return _(u"Game #{0} created at {1} (site {2})").format(
            self.game_id,
            timezone.localtime(self.created).strftime(u"%Y-%m-%d %H:%M"),
            self.site)

    def hard_expiry(self):
        if GAME_HARD_TIMEOUT is not None:
            return self.created + timezone.timedelta(0, GAME_HARD_TIMEOUT * 60)

    def soft_expiry(self):
        if GAME_SOFT_TIMEOUT is not None:
            return self.last_used + timezone.timedelta(
                0, GAME_SOFT_TIMEOUT * 60)

    def end_time(self):
        hard_expiry = self.hard_expiry()
        end_time = get_endtime()
        if hard_expiry is not None and end_time is not None:
            return min(end_time, hard_expiry)
        elif hard_expiry is None:
            # may be None
            return end_time
        else:
            # may be None
            return hard_expiry

    def is_expired(self):
        # game expired, because no one used it
        now = timezone.now()
        diff_last_used = now - self.last_used
        diff_created = now - self.created

        seconds_since_last_used = \
            diff_last_used.days * 24 * 60 * 60 + diff_last_used.seconds
        seconds_since_created = \
            diff_created.days * 24 * 60 * 60 + diff_created.seconds

        # game expired, because nobody used it
        if GAME_SOFT_TIMEOUT and seconds_since_last_used \
                > (GAME_SOFT_TIMEOUT * 60):
            return True

        # game expired, because its too old, even when someone is using it
        elif GAME_HARD_TIMEOUT and seconds_since_created \
                > (GAME_HARD_TIMEOUT * 60):
            return True
        else:
            return False

    def num_users(self):
        return self.bingoboard_set.count()

    def num_active_users(self):
        if USER_ACTIVE_TIMEOUT:
            return self.bingoboard_set.exclude(
                last_used__lte=timezone.now()
                - timezone.timedelta(0, 60 * USER_ACTIVE_TIMEOUT)
            ).count()
        else:
            return self.num_users()

    def words_with_votes(self, only_topics=True):
        """
            returns a list with words ordered by the number of votes
            annotated with the number of votes in the "votes" property.
        """
        result = Word.objects.filter(
            bingofield__board__game__id=self.id).exclude(
            type=WORD_TYPE_MIDDLE)

        if only_topics:
            result = result.exclude(bingofield__word__type=WORD_TYPE_META)

        result = result.annotate(
            votes=Sum("bingofield__vote")).order_by("-votes").values()

        for item in result:
            item['votes'] = max(0, item['votes'])
            if result[0]['votes'] != 0:
                item['percent'] = float(item['votes']) / result[0]['votes'] * 100
            else:
                item['percent'] = 0
        return result

    def all_words_with_votes(self):
        return self.words_with_votes(only_topics=False)

    def rating(self):
        return self.bingoboard_set.exclude(rating=None).aggregate(
            rating=models.Avg('rating'))['rating'] or 0

    def num_ratings(self):
        return self.bingoboard_set.exclude(rating=None).aggregate(
            num=models.Count('rating'))['num']

    def save(self):
        if self.pk is None:
            games = Game.objects.filter(site=self.site)
            current_id = games.aggregate(
                max_id=models.Max('game_id'))['max_id']
            if current_id is None:
                self.game_id = 1
            else:
                self.game_id = current_id + 1
        return super(Game, self).save()


def _get_random_words(site):
    all_words = Word.objects.filter(site=site).order_by("?")
    middle_words = all_words.filter(site=site, type=WORD_TYPE_MIDDLE).order_by("?")
    words = all_words.exclude(type=WORD_TYPE_MIDDLE)
    if middle_words.count() == 0:
        raise ValidationError(_(u"No middle words in database"))
    middle = middle_words[0]
    if words.count() < 24:
        raise ValidationError(_(u"Not enough (non-middle) words in database"))
    return list(words), middle


RATINGS = (
    (None, _("unrated")),
    (1, _("{0} stars").format(1)),
    (2, _("{0} stars").format(2)),
    (3, _("{0} stars").format(3)),
    (4, _("{0} stars").format(4)),
    (5, _("{0} stars").format(5))
)


class BingoBoard(models.Model):
    board_id = models.IntegerField(blank=True, null=True)
    game = models.ForeignKey("Game")
    color = RGBColorField()
    ip = models.GenericIPAddressField(blank=True, null=True)
    user = models.ForeignKey(User, blank=True, null=True)
    password = models.CharField(max_length=255, blank=True, null=True)
    created = models.DateTimeField(auto_now_add=True)
    last_used = models.DateTimeField(auto_now_add=True)
    rating = models.IntegerField(choices=RATINGS, null=True, blank=True,
        default=None)

    class Meta:
        ordering = ("-board_id",)
        # board_id, game__site is not possible
        unique_together = ("board_id", "game")

    def get_absolute_url(self):
        return reverse('bingo.views.bingo', kwargs={"board_id": self.board_id})

    def clean(self):
        try:
            board = BingoBoard.objects.get(
                board_id=self.board_id, game__site=self.game.site)
            # if its not an edit of the same BingoBoard
            if board.pk != self.pk:
                raise ValidationError(
                    _(u"BingoBoard ID not unique for site {0}").format(
                        self.game.site))
        except BingoBoard.DoesNotExist:
            return  # everything okay

    def save(self):
        if self.ip is None and self.user is None:
            raise ValidationError(
                _(u"BingoBoard must have either an ip or an user"))

        if self.pk is None:
            # unique_together for optional fields
            game_boards = BingoBoard.objects.filter(game=self.game)
            if not self.user is None:
                if game_boards.filter(user=self.user).count() > 0:
                    raise ValidationError(
                        _(u"game and user must be unique_together"))
            # ip only needs to be unique for anonymous users
            # to check if the ip is relevant:
            # - check the current board does not have an user
            # - filter for existing boards from this ip only with user=None
            if not self.ip is None and not self.user:
                if game_boards.filter(ip=self.ip, user=None).count() > 0:
                    raise ValidationError(
                        _(u"game and ip must be unique_together"))

            # generate a color
            self.color = "#%x%x%x" % (
                randint(COLOR_FROM, COLOR_TO),
                randint(COLOR_FROM, COLOR_TO),
                randint(COLOR_FROM, COLOR_TO))

            # first create the fields, so the board will
            # not be saved, when field creation fails
            fields = self.create_bingofields()

            # create the board
            super(BingoBoard, self).save()

            # now that the board has a pk, save the fields
            for field in fields:
                field.board = self
                field.save()

            # then create a board_id as atomic as possible for two fields,
            # which need to be unique together
            bingo_boards = BingoBoard.objects.filter(
                game__site=self.game.site)
            current_id = bingo_boards.aggregate(
                max_id=models.Max('board_id'))['max_id']
            new_id = current_id + 1 if current_id else 1
            BingoBoard.objects.filter(id=self.id).update(
                board_id=new_id)

            # set the id for the current instance
            self.board_id = new_id

        else:
            super(BingoBoard, self).save()

    def create_bingofields(self):
        count = 0
        words, middle = _get_random_words(site=self.game.site)
        fields = []
        for i in xrange(25):
            # 13th field = middle
            if i == 12:
                fields.append(BingoField(word=middle, position=i+1))
            else:
                fields.append(BingoField(word=words[count], position=i+1))
                count += 1
        # create fields without position for every
        # active word, which is not on the board, too.
        # the index is 24, because there are only 24
        # non-middle words on the field
        assert count == 24
        for word in words[24:]:
            fields.append(BingoField(word=word, position=None))

        return fields

    def get_board_fields(self):
        return self.bingofield_set.exclude(position=None).order_by("position")

    def get_all_word_fields(self):
        return self.bingofield_set.exclude(word__type=WORD_TYPE_MIDDLE)

    def get_middle_field(self):
        return self.bingofield_set.filter(word__type=WORD_TYPE_MIDDLE)[0]

    def get_created(self):
        """
            get created field with BINGO_IMAGE_DATETIME_FORMAT formatting
        """
        return timezone.localtime(self.created).strftime(
            BINGO_IMAGE_DATETIME_FORMAT)

    def get_last_used(self):
        """
            get last_used field with BINGO_IMAGE_DATETIME_FORMAT formatting
        """
        return timezone.localtime(self.last_used).strftime(
            BINGO_IMAGE_DATETIME_FORMAT)

    def thumbnails_enabled(self):
        return THUMBNAILS_ENABLED

    def __unicode__(self):
        return _(u"BingoBoard #{0} created by {1} (site {2})").format(
            self.board_id,
            self.user if self.user else self.ip,
            self.game.site)


def position_validator(value):
    if not value is None and not (0 < value < 26):
        raise ValidationError(_(
            _(u"invalid position. valid values range: 1-25 or None")))


class BingoField(models.Model):
    word = models.ForeignKey("Word")
    board = models.ForeignKey("BingoBoard")
    position = models.SmallIntegerField(
        validators=[position_validator],
        blank=True, null=True, default=None)
    vote = models.SmallIntegerField(default=0)

    class Meta:
        ordering = ("board", "-position")

    def is_middle(self):
        return self.position == 13

    def num_votes(self):
        if self.word.type == WORD_TYPE_MIDDLE:
            return 0
        vote_counts_cachename = 'vote_counts_game={0:d}'.format(
            self.board.game.id)
        vote_counts = cache.get(vote_counts_cachename, None)
        if not vote_counts or not self.word.id in vote_counts:
            votes = Word.objects.exclude(type=WORD_TYPE_MIDDLE).filter(
                bingofield__board__game=self.board.game).annotate(
                votes=Sum('bingofield__vote')).values("id", "votes")

            vote_counts = {}
            for vote in votes:
                vote_counts[vote['id']] = vote['votes']
            cache.set(vote_counts_cachename, vote_counts)
        return max(0, vote_counts.get(self.word.id, 0))

    def clean(self):
        if self.is_middle() and self.word.type != WORD_TYPE_MIDDLE:
            raise ValidationError(_(
                u"The BingoField has middle position, "
                u"but the word is no middle word"))
        elif not self.is_middle() and self.word.type == WORD_TYPE_MIDDLE:
            raise ValidationError(_(
                u"The BingoField is not in the middle, "
                u"but the word is a middle word"))

    def __unicode__(self):
        if self.position is not None:
            return _(u"BingoField: word={0}, pos=({1},{2}){3})").format(
                self.word, self.position/5+1, self.position % 5,
                _(u" [middle]") if self.is_middle() else u"")
        else:
            return _(u"BingoField: word={0}, (not on the board)").format(
                self.word)
